File: java/src/org/raml/amf/core/domain/DomainModel.java
package org.raml.amf.core.domain;

import api_modeling_framework.model.document.Node;
import api_modeling_framework.model.document.DocumentSourceMap;
import api_modeling_framework.model.domain.*;
import org.raml.amf.core.exceptions.InvalidModelException;
import org.raml.amf.core.Model;
import org.raml.amf.utils.Clojure;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by antoniogarrote on 04/05/2017.
 */
public class DomainModel extends Model {

    static {
        Clojure.require(Clojure.API_MODELING_FRAMEWORK_MODEL_DOMAIN);
    }

    public DomainModel(Object rawModel) {
        super(rawModel);
    }

    @Override
    public Object clojureModel() {
        return this.rawModel;
    }

    public static DomainModel fromRawModel(Object rawModel) {
        if (rawModel instanceof ParsedAPIDocumentation) {
            return new APIDocumentation((ParsedAPIDocumentation) rawModel);
        } else if (rawModel instanceof ParsedEndPoint) {
            return new EndPoint((ParsedEndPoint) rawModel);
        } else if (rawModel instanceof ParsedOperation) {
            return new Operation((ParsedOperation) rawModel);
        } else if (rawModel instanceof ParsedRequest) {
            return new Request((ParsedRequest) rawModel);
        } else if (rawModel instanceof ParsedResponse) {
            return new Response((ParsedResponse) rawModel);
        } else if (rawModel instanceof Payload) {
            return new Payload((ParsedPayload) rawModel);
        } else if (rawModel instanceof ParsedType) {
            return new Type((ParsedType) rawModel);
        } else {
            throw new InvalidModelException(new Exception("Unknown DomainModel class " + rawModel));
        }
    }

    public String getId() {
        return (String) wrappedNode().id();
    }

    public String getName() {
        return (String) wrappedNode().name();
    }

    public void setName(String name) {
        this.rawModel = Clojure.setKw(this.rawModel, "name", name);
    }

    public Boolean getAbstract() {
        return (Boolean) Clojure.getKw(this.rawModel, "abstract") || false;
    }

    public void setAbstract(Boolean abstractBool) {
        this.rawModel = Clojure.setKw(this.rawModel, "abstract", abstractBool);
    }

    public String getDescription() {
        return (String) wrappedNode().name();
    }

    public void setDescription(String description) {
        this.rawModel = Clojure.setKw(this.rawModel, "description", description);
    }

    public List<SourceMap> getSourceMaps() {
        List operations =  (List) this.wrappedNode().sources();
        List<DocumentSourceMap> tmp = Clojure.toJavaList(operations);
        ArrayList<SourceMap> eps = new ArrayList<>();
        for(DocumentSourceMap x : tmp) {
            SourceMap parsed = new SourceMap(x);
            eps.add(parsed);
        }

        return eps;
    }

    public void setSourceMaps(List<SourceMap> sourceMaps) {
        ArrayList<Object> raws = new ArrayList<>();
        for(SourceMap x : sourceMaps) {
            raws.add(x.clojureModel());
        }

        this.rawModel = Clojure.setKw(this.rawModel, "sources", Clojure.list(raws));
    }

    public List<DomainModel> getExtends() {
        List operations =  (List) Clojure.getKw(this.rawModel, "extends");
        List<Object> tmp = Clojure.toJavaList(operations);
        ArrayList<DomainModel> eps = new ArrayList<>();
        for(Object x : tmp) {
            eps.add(DomainModel.fromRawModel(x));
        }

        return eps;
    }

    public void setExtends(List<DomainModel> toExtend) {
        ArrayList<Object> raws = new ArrayList<>();
        for(DomainModel x : toExtend) {
            raws.add(x.clojureModel());
        }

        this.rawModel = Clojure.setKw(this.rawModel, "extends", Clojure.list(raws));
    }

    protected void setId(String id) {
        this.rawModel = Clojure.setKw(this.rawModel, "id", id);
    }
    protected Node wrappedNode() {
        return (Node) this.rawModel;
    }

    public String toString() {
        return (this.getId() + " :: " + super.toString());
    }
}

File: bindings/js/src/core/domain/DomainModel.ts
/**
 * Created by antoniogarrote on 05/05/2017.
 */

import {Clojure} from "../../Clojure";
import {Model} from "../Model";

/**
 * Base class for all Domain Model elements
 */
export class DomainModel extends Model {

    constructor(protected rawModel: any) {
        super(rawModel);
    }

    /**
     * Factory method building the right wrapper Java DomainModel subclass for the provided Domain Model Clojure data structure
     * @param rawModel native Clojure encoded model
     * @return The right DocumentModel
     */
    public static fromRawModel(rawModel: any): DomainModel {
        const type  = Clojure.core.type(rawModel);
        if (type === Clojure.amf_domain.ParsedAPIDocumentation) {
            return new APIDocumentation(rawModel);
        } else if (type === Clojure.amf_domain.ParsedEndPoint) {
                return new EndPoint(rawModel);
        } else if (type === Clojure.amf_domain.ParsedOperation) {
            return new Operation(rawModel);
        } else if (type === Clojure.amf_domain.ParsedRequest) {
            return new Request(rawModel);
        } else if (type === Clojure.amf_domain.ParsedResponse) {
            return new Response(rawModel);
        } else if (type === Clojure.amf_domain.ParsedPayload) {
            return new Payload(rawModel);
        } else if (type === Clojure.amf_domain.ParsedType) {
            return new Type(rawModel);
        } else {
            throw new Error(`Unknown model ${type}`)
        }
    }

    public clojureModel(): any {
        return this.rawModel;
    }

    /**
     * Returns the unique URI identifier the Domain Element in the AMF graph
     * @return
     */
    public getId(): string {
        return Clojure.amf_document.id(this.rawModel);
    }

    /**
     * Returns a human readable string identifying the Domain Element
     * @return
     */
    public getName(): string | undefined {
        return Clojure.amf_domain.name(this.rawModel);
    }

    public setName(name: string | undefined) {
        this.rawModel = Clojure.amf.update(DomainModel.domain_builder, this.rawModel, Clojure.kw("name"), name);
    }

    /**
     * If true, the Domain Element is abstract and should be used to be extended by other Domain Elements
     * @return
     */
    public getAbstract(): boolean | undefined {
        return Clojure.amf_domain.abstract(this.rawModel);
    }

    public setAbstract(abstract: boolean | undefined) {
        this.rawModel = Clojure.amf.update(DomainModel.domain_builder, this.rawModel, Clojure.kw("abstract"), abstract);
    }

    /**
     * List of Domain Elements this element extends
     * @return
     */
    public getExtends(): DomainModel[] {
        let extensions = Clojure.amf_domain.extends(this.rawModel);
        return Clojure.cljsMap(extensions, (e) => DomainModel.fromRawModel(e));
    }

    public setExtensions(extensions: DomainModel[]) {
        const newExtensions = extensions.map(e => e.clojureModel());
        this.rawModel = Clojure.amf.update(DomainModel.domain_builder, this.rawModel, Clojure.kw("extends"), Clojure.toCljsSeq(newExtensions));
    }

    /**
     * Optional list of source maps with lexical information associated to the Domain Element
     * @return
     */
    public getSourceMaps(): SourceMap[] {
        let sourceMaps = Clojure.amf_domain.sources(this.rawModel);
        return Clojure.cljsMap(sourceMaps, (e) => new SourceMap(e));
    }

    public setSourceMaps(sourceMaps: SourceMap[]) {
        const newSourceMaps = sourceMaps.map(e => e.clojureModel());
        this.rawModel = Clojure.amf.update(DomainModel.domain_builder, this.rawModel, Clojure.kw("sources"), Clojure.toCljsSeq(newSourceMaps));
    }

    public static domain_builder = Clojure.amf.map__GT_JSDomainBuilder()
}

import {APIDocumentation} from "./APIDocumentation";
import {EndPoint} from "./EndPoint";
import {Operation} from "./Operation";import { SourceMap } from "./SourceMap";
import {Payload} from "./Payload";
import {Parameter} from "./Parameter";
import {Type} from "./Type";
import {Request} from "./Request";
import {Response} from "./Response";
